contextualize_pnid_files.ipynb

```diff
@@ -0,0 +1,207 @@
+   "outputs": [],
+   "source": [
+    "# Get files with source\n",
+    "source = \"Pearl: ASU 1 Unit 12100 PEFS\"\n",
+    "files = client.files.list(source=source, limit=-1)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create entites based on asset names\n",
+    "entities = [asset.name for asset in assets]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Parse files, create SVGs with the found assets, contextualize the original files in CDF, upload the new SVGs to CDF\n",
+    "\n",
+    "file_metadata_updates = list()\n",
+    "\n",
+    "# add source for the new SVGs\n",
+    "svg_source=\"contextualization\"\n",
+    "\n",
+    "# add mime_type\n",
+    "svg_mime_type = \"image/svg+xml\"\n",
+    "\n",
+    "# overwrite the svg file if already uploaded\n",
+    "svg_overwrite = True\n",
+    "\n",
+    "for file in files:\n",
+    "    print(f\"Parse and convert P&ID to SVG, input file: {file.name}\")\n",
+    "\n",
+    "    # run pnid_parse job\n",
+    "    job = await client.pnid_parsing.parse(file_id=file.id, entities=entities, partial_match=True)\n",
+    "    \n",
+    "    # store the svg url with the highlighted detected entities\n",
+    "    svg_url = job.args[\"svg_url\"]\n",
+    "    \n",
+    "    # store the results with the entities found and the corresponding bounding boxes \n",
+    "    results = job.result\n",
+    "\n",
+    "    # get only entities from the results\n",
+    "    entities_found = [result[\"text\"] for result in results]\n",
+    "    if not entities_found:\n",
+    "        # skip to the next file if no assets are found\n",
+    "        continue\n",
+    "    \n",
+    "    # remove duplicate entities\n",
+    "    entities_found = list(set(entities_found))\n",
+    "    \n",
+    "    # make a list of asset_ids based on the found entities \n",
+    "    asset_ids_found = set()\n",
+    "    for entity in entities_found:\n",
+    "        # asset name could correspond to several asset_ids\n",
+    "        asset_ids_found.update(asset_id_dict[entity])\n",
+    "        \n",
+    "    asset_names      = ','.join(map(str, entities_found))\n",
+    "    print(f\"\\tFound assets: {asset_names}\")\n",
+    "    \n",
+    "    # update metadata for the original file\n",
+    "    file_metadata_update = FileMetadataUpdate(id=file.id).asset_ids.set(list(asset_ids_found))\n",
+    "    file_metadata_updates.append(file_metadata_update)\n",
+    "    \n",
+    "    # add metadata to the new svg, e.g.,\n",
+    "    svg_metadata = {\"OriginalSource\": file.source}\n",
+    "    \n",
+    "    # keep the name for the file, replacing the extention\n",
+    "    name_svg = os.path.splitext(file.name)[0] + \".svg\"\n",
+    "    \n",
+    "    # give a meaningful external_id\n",
+    "    external_id_svg = file.external_id + \"+svg\"\n",
```

Comment:
What if the file does not have an external id? Maybe it's ok as this is just a template.

Reply:
I think it's fine with external_id, application teams argument for external_ids so OK in the template. The user of the template could delete it if they want when copying.

Reply:
I agree!